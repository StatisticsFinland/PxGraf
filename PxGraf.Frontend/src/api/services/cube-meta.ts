/* istanbul ignore file */

import ApiClient from "api/client";
import { useQuery } from "react-query";
import { ICubeMeta } from "types/cubeMeta";
import { defaultQueryOptions } from "utils/ApiHelpers";

/**
 * Interface for cube metadata result. Cube is a multi-dimensional result of Px file data filtered by a query.
 * @property {boolean} isLoading - Flag to indicate if the data is still loading.
 * @property {boolean} isError - Flag to indicate if an error occurred during loading.
 * @property {ICubeMeta} data - The cube meta data as an @see {@link ICubeMeta} object.
 */
export interface ICubeMetaResult {
    isLoading: boolean;
    isError: boolean;
    data: ICubeMeta;
}

const fetchCubeMeta = async(idStack: string[]): Promise<ICubeMeta> => {
    const client = new ApiClient();
    const url = `creation/cube-meta/${idStack.join('/')}`;

    return await client.getAsync(url);
}

export const useCubeMetaQuery = (idStack: string[]): ICubeMetaResult => {
    return useQuery(
        ['cube-meta', idStack],
        () => fetchCubeMeta(idStack),
        defaultQueryOptions
    );
}