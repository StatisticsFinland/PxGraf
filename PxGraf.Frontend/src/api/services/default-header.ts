/* istanbul ignore file */

import ApiClient from "api/client";
import { merge } from "lodash";
import { useQuery } from "react-query";
import { MultiLanguageString } from "types/multiLanguageString";
import { Query } from "types/query";
import { buildTableReference, defaultQueryOptions } from "utils/ApiHelpers";

/**
 * Interface for the chart's header result
 * @property {boolean} isLoading - Flag to indicate if the data is still loading.
 * @property {boolean} isError - Flag to indicate if an error occurred while retrieving the header.
 * @property {MultiLanguageString} data - The header data represented as a multi-language string.
 */
export interface IHeaderResult {
    isLoading: boolean;
    isError: boolean;
    data: MultiLanguageString;
}

const fetchDefaultHeader = async(idStack: string[], query: Query): Promise<MultiLanguageString> => {
    const client = new ApiClient();
    const requestBody = JSON.stringify(
        merge(
            { tableReference: buildTableReference(idStack) },
            { variableQueries: query }
        )
    );
  
    const url = 'creation/default-header';
    return await client.postAsync(url, requestBody);
}

export const useDefaultHeaderQuery = (idStack: string[], query: Query): IHeaderResult => {
    return useQuery(
        ['default-header', ...idStack, query],
        () => fetchDefaultHeader(idStack, query),
        {
            ...defaultQueryOptions,
            enabled: query != null,
        }
    );
}