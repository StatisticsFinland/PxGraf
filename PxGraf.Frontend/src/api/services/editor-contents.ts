/* istanbul ignore file */

import ApiClient from "api/client";
import { useQuery } from "react-query";
import { IEditorContentsResponse } from "types/editorContentsResponse";
import { buildCubeQuery, defaultQueryOptions, useDebounceState } from "utils/ApiHelpers";
import { ICubeQuery, Query } from "types/query";

/**
 * Interface for editor contents result. Editor contents are used by the editor to populate the settings for adjusting the visualization.
 * @property {boolean} isLoading - Flag to indicate if the data is still loading.
 * @property {boolean} isError - Flag to indicate if an error occurred during loading.
 * @property {IEditorContentsResponse} data - The editor contents as an @see {@link IEditorContentsResponse} object.
 */ 
export interface IEditorContentsResult {
    isLoading: boolean;
    isError: boolean;
    data: IEditorContentsResponse;
}

const fetchEditorContents = async (idStack: string[], query: Query, metaEdits: ICubeQuery): Promise<IEditorContentsResponse> => {
    const client = new ApiClient();
    const url = 'creation/editor-contents';
    const requestBody = JSON.stringify(buildCubeQuery(query, metaEdits, idStack));
    return await client.postAsync(url, requestBody);
}

export const useEditorContentsQuery = (idStack: string[], query: Query, cubeQuery: ICubeQuery): IEditorContentsResult => {
    [idStack, query, cubeQuery] = useDebounceState(1000, idStack, query, cubeQuery);
    const queryKey = ['editor-contents', ...idStack, query, cubeQuery];
    return useQuery(
        queryKey,
        () => fetchEditorContents(idStack, query, cubeQuery),
        {
            ...defaultQueryOptions,
            enabled: query != null,
            keepPreviousData: true,
        }
    );
}