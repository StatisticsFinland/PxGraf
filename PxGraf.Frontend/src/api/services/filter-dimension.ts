/* istanbul ignore file */

import ApiClient from "api/client";
import { useQuery } from "react-query";
import { IValueFilter, Query } from "types/query";
import { buildTableReference, defaultQueryOptions } from "utils/ApiHelpers";

/**
 * Interface for variable filter result
 * @property {boolean} isLoading - Flag to indicate if the data is still loading.
 * @property {boolean} isError - Flag to indicate if an error occurred during loading.
 * @property {{ [key: string]: string[] }} data - The variable filter data represented by a dictionary. Key is the variable code and value is a list of strings that represent value codes for that variable.
 */
export interface IFilterVariableResult {
    isLoading: boolean;
    isError: boolean;
    data: {[key: string]: string[]}
}

const fetchResolveVariableFilter = async (idStack: string[], variableFilters: { [key: string]: IValueFilter }): Promise<{ [key: string]: string[] }> => {
    const client = new ApiClient();
    const requestBody = JSON.stringify(
        {
            tableReference: buildTableReference(idStack),
            filters: variableFilters,
        }
    );

    const url = 'creation/filter-dimension'
    return await client.postAsync(url, requestBody);
}

export const useResolveVariableFiltersQuery = (idStack: string[], query: Query): IFilterVariableResult => {
    const varQueries = Object.entries(query ?? {});
    const varFilters =  Object.fromEntries(varQueries.map(([variableCode, variableQuery]) => {
        return [variableCode, variableQuery.valueFilter]
    }));

    return useQuery(
        ['filter-dimension', ...idStack, query],
        () => fetchResolveVariableFilter(idStack, varFilters),
        defaultQueryOptions
    );
}