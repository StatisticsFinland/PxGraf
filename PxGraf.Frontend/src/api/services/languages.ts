/* istanbul ignore file */

import ApiClient from "api/client";
import { useQuery } from "react-query";
import { defaultQueryOptions } from "utils/ApiHelpers";

const fetchLanguages = async (): Promise<string[]> => {
    const client = new ApiClient();
    const url = "creation/languages/";

    return await client.getAsync(url);
};

/**
 * Parses a list of languages into a string for display.
 * @param {string[]} languages List of languages.
 * @returns All listed languages in one string, in upper case.
 */
export const parseLanguageString = (languages: string[]): string => {
    return `(${languages.join(", ").toUpperCase()})`;
}

/**
 * Chooses a listing language for the given listing item, trying to prefer the selected uiLanguage.
 * @param {string} id Id of the database or table.
 * @param {string[]} languages Languages supported by the listing item.
 * @param {string} uiLanguage Currently selected UI language.
 * @returns UI language, if it is supported by the listing item's languages. Otherwise the default language of the item.
 */
export const getDatabaseListingLanguage = (id: string, languages: string[], uiLanguage: string): string => {
    return languages.includes(uiLanguage) ? uiLanguage : languages[0];
}

export const useLanguagesQuery = (): string[] => {
    const queryResult = useQuery(
        ['languages'],
        () => fetchLanguages(),
        defaultQueryOptions
    );
    return queryResult.data;
};