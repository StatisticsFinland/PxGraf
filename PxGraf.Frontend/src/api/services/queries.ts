/* istanbul ignore file */

import ApiClient from "api/client";
import { useMutation, useQuery } from "react-query";
import { ICubeQuery, Query } from "types/query";
import { IVisualizationSettings } from "types/visualizationSettings";
import { buildCubeQuery, defaultQueryOptions } from "utils/ApiHelpers";

/**
 * Response object for saving a query.
 * @property {string} id - The id of the saved query.
 */
export interface ISaveQueryResponse {
    id: string;
}

/**
 * Response object for fetching a saved query.
 * @property query - The query object consisting of the table reference and variable queries @see {@link Query}.
 * @property {IVisualizationSettings} settings - The visualization settings for the query.
 */
export interface IFetchSavedQueryResponse {
    query: {
        tableReference: {
            name: string;
            hierarchy: string[];
        };
        variableQueries: Query;
    } & ICubeQuery;
    settings: IVisualizationSettings;
}

/**
 * Interface for save query result
 * @property {boolean} isLoading - Flag to indicate if the data is still loading.
 * @property {boolean} isError - Flag to indicate if an error occurred during loading.
 * @property {boolean} isSuccess - Flag to indicate if the request was successful.
 * @property {ISaveQueryResponse} data - The save query response data.
 * @property {function} mutate - Function to mutate the state of the query.
 */
export interface ISaveQueryResult {
    isLoading: boolean;
    isError: boolean;
    isSuccess: boolean;
    data: ISaveQueryResponse;
    mutate: (property: boolean) => void;
}

/**
 * Interface for fetch saved query result
 * @property {boolean} isLoading - Flag to indicate if the data is still loading.
 * @property {boolean} isError - Flag to indicate if an error occurred during loading.
 * @property {boolean} isSuccess - Flag to indicate if the request was successful.
 * @property {IFetchSavedQueryResponse} data - The fetch saved query response data.
 */
export interface IFetchSavedQueryResult {
    isLoading: boolean;
    isError: boolean;
    isSuccess: boolean;
    data: IFetchSavedQueryResponse;
}

const sendSaveRequest = async (
    idStack: string[],
    query: Query,
    metaEdits: ICubeQuery,
    selectedVisualization: string,
    visualizationSettings: IVisualizationSettings,
    archive: boolean | void
): Promise<ISaveQueryResponse> => {

    const client = new ApiClient();

    const requestBody = JSON.stringify({
        query: buildCubeQuery(query, metaEdits, idStack),
        settings: {
            ...visualizationSettings,
            selectedVisualization: selectedVisualization,
        }
    });
    const url = archive ? 'sq/archive' : 'sq/save';
    return await client.postAsync(url, requestBody);
}

export const fetchSavedQuery = async (queryId: string): Promise<IFetchSavedQueryResponse> => {
    const client = new ApiClient();
    const url = 'sq/' + queryId;
    const response = await client.getAsync(url);

    // Check for VariableQueries and assign to variableQueries
    if (!response.query?.variableQueries && response.query?.VariableQueries) {
        response.query.variableQueries = response.query.VariableQueries;
    }

    return response as IFetchSavedQueryResponse;
};

export const useFetchSavedQuery = (queryId: string): IFetchSavedQueryResult => {
    return useQuery(
        [queryId],
        () => fetchSavedQuery(queryId),
        defaultQueryOptions
    );
};

export const useSaveMutation = (idStack: string[], query: Query, metaEdits: ICubeQuery, selectedVisualization: string, visualizationSettings: IVisualizationSettings): ISaveQueryResult => {
    return useMutation(
        (archive) => sendSaveRequest(idStack, query, metaEdits, selectedVisualization, visualizationSettings, archive),
        { retry: false }
    );
}