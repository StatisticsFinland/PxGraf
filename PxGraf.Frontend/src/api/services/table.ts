/* istanbul ignore file */

import ApiClient from "api/client";
import { useQuery } from "react-query";
import { defaultQueryOptions } from "utils/ApiHelpers";
import { MultiLanguageString } from "../../types/multiLanguageString";

/**
 * Interface for a table list response.
 * @property {string} id - The id of the table.
 * @property {string} type - The type of the table where 't' is a table and 'l' is a list item.
 * @property {string} updated - The last updated date of the table.
 * @property {MultiLanguageString} text - The description of the table as a multi-language string.
 * @property {string[]} languages - Available languages of the table.
 */
export interface ITableListResponse {
    id: string;
    type?: 't' | 'l';
    updated?: string;
    text: MultiLanguageString;
    languages?: string[];
}

/**
 * Interface for a table result.
 * @property {boolean} isLoading - Flag to indicate if the data is still loading.
 * @property {boolean} isError - Flag to indicate if an error occurred during loading.
 * @property {ITableListResponse[]} data - The table list response data.
 */
export interface ITableResult {
    isLoading: boolean;
    isError: boolean;
    data: ITableListResponse[];
}

const fetchTable = async (idStack: string[], lang: string): Promise<ITableListResponse[]> => {
    const client = new ApiClient();
    const path = idStack.join("/");
    const url = "creation/data-bases/" + path;
    const getParams = {
        'lang': lang
    }

    return await client.getAsync(url, getParams);
};

export const sortTableData = (data: ITableListResponse[], primaryLanguage: string, databaseLanguages: string[]): ITableListResponse[] => {
    return [...data].sort((a, b) => {
        const textA = a.text[primaryLanguage] || a.text[databaseLanguages[0]];
        const textB = b.text[primaryLanguage] || b.text[databaseLanguages[0]];

        return textA.localeCompare(textB, primaryLanguage);
    });
};

export const useTableQuery = (idStack: string[], lang: string): ITableResult => {
    // If trying to fetch a database without a language, return empty
    const fetchFunction = lang ?
        () => fetchTable(idStack, lang) :
        () => Promise.resolve([]);

    return useQuery(
        ['table', ...idStack, lang],
        fetchFunction,
        defaultQueryOptions,
    );
};