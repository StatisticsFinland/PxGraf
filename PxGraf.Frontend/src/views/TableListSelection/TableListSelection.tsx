import { useParams } from "react-router-dom";
import { useTranslation } from 'react-i18next';
import { List, ListItem, Divider, Container, Skeleton, Alert } from '@mui/material';

import { UiLanguageContext } from 'contexts/uiLanguageContext';
import React from "react";
import { DirectoryInfo } from 'components/DirectoryInfo/DirectoryInfo';
import TableInfo from "components/TableInfo/TableInfo";
import styled from "styled-components";
import { useTableQuery } from "api/services/table";
import { useLanguagesQuery } from "../../api/services/languages";
import { sortTableData } from 'utils/sortingHelpers';

const TableQueryAlert = styled(Alert)`
  width: 100%;
`;

const StyledSkeleton = styled(Skeleton)`
  height: 128px;
  width: 100%;
`;

const TableListSelectionWrapper = styled(Container)`
  padding: 8px;
`;

/**
 * Component for displaying details of a table list subfolder's contents.
 */
export const TableListSelection: React.FC = () => {
    const { t } = useTranslation();
    const { language } = React.useContext(UiLanguageContext);
    const databaseLanguages: string[] = useLanguagesQuery();
    let primaryLanguage: string;
    if (databaseLanguages) {
        primaryLanguage = databaseLanguages.includes(language) ? language : databaseLanguages[0];
    }

    const params = useParams();
    const path = params["*"].split("/").filter(p => p.length > 0);
    const { isLoading, isError, data } = useTableQuery(path, primaryLanguage);

    React.useEffect(() => {
        document.title = `${t("pages.tableList")} | PxGraf`;
    }, []);

    const sortedData = React.useMemo(() => {
        if (!data) return null;

        return sortTableData(data, primaryLanguage);
    }, [data, primaryLanguage, databaseLanguages]);

    let content: React.ReactNode;
    if (isError) {
        content =
            <>
                <ListItem>
                    <TableQueryAlert id="mainContent" severity="error">{t("error.contentLoad")}</TableQueryAlert>
                </ListItem>
                <Divider />
            </>

    }
    else if (isLoading) {
        content =
            <>
                <ListItem>
                    <StyledSkeleton variant="rectangular" />
                </ListItem>
                <Divider />
                <ListItem>
                    <StyledSkeleton variant="rectangular" />
                </ListItem>
                <Divider />
                <ListItem>
                    <StyledSkeleton variant="rectangular" />
                </ListItem>
                <Divider />
            </>
    }
    else {
        content = sortedData.map((item) =>
            item.type === "t" ?
                <TableInfo path={params["*"]} item={item} key={`${item.id}-table-info`} /> :
                <DirectoryInfo path={params["*"]} item={item} key={`${item.id}-directory-info`} />
        );
    }

    // actual rendered component
    return (
        <TableListSelectionWrapper maxWidth="md">
            <List sx={{ bgcolor: 'background.paper' }}>
                {content}
            </List>
        </TableListSelectionWrapper>
    );
}

export default TableListSelection;