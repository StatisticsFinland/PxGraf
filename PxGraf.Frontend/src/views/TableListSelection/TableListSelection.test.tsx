import React from 'react';
import { render } from '@testing-library/react';
import { ITableResult } from 'api/services/table';
import UiLanguageContext from 'contexts/uiLanguageContext';
import '@testing-library/jest-dom';
import { MemoryRouter } from 'react-router-dom';
import TableListSelection from './TableListSelection';

jest.mock('react-i18next', () => ({
    ...jest.requireActual('react-i18next'),
    useTranslation: () => {
        return {
            t: (str: string) => str,
            i18n: {
                changeLanguage: () => new Promise(() => {}),
            },
        };
    },
}));

jest.mock('envVars', () => ({
    PxGrafUrl: 'pxGrafUrl.fi/',
    PublicUrl: 'publicUrl.fi/'
}));

jest.mock('api/services/languages', () => ({
    ...jest.requireActual('api/services/languages'),
    useLanguagesQuery: () => {
        return {
            data: ['fi', 'en', 'sv']
        }
    },
}));

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useParams: () => {
        return {
            '*': 'foo/bar'
        };
    },
}));

const setLanguage = jest.fn();
const language = 'fi';

const setLanguageTab = jest.fn();
const languageTab = 'fi';

const availableUiLanguages = ['fi', 'en', 'sv'];

const uiContentLanguage = "fi";
const setUiContentLanguage = jest.fn();

const mockTableResult: ITableResult = {
    isLoading: false,
    isError: false,
    data:
    {
        headers: [
            {
                name: { 'fi': 'foo1-fi', 'en': 'foo1-en', 'sv': 'foo1-sv' },
                code: 'id1',
                languages: ['fi', 'en', 'sv']
            },
            {
                name: { 'fi': 'foo2-fi', 'en': 'foo2-en' },
                code: 'id2',
                languages: ['fi', 'en']
            }
        ],
        files: [
            {
                name: { 'fi': 'foo3-fi', 'en': 'foo3-en', 'sv': 'foo3-sv' },
                code: 'id2',
                lastUpdated: '1.1.2000',
                languages: ['fi', 'en', 'sv']
            }
        ],
    }
}

const mockLanguagesResult = ["fi", "en", "sv"];


jest.mock('api/services/languages', () => ({
    ...jest.requireActual('api/services/languages'),
    useLanguagesQuery: () => {
        return mockLanguagesResult;
    },
}));

jest.mock('api/services/table', () => ({
    ...jest.requireActual('api/services/table'),
    useTableQuery: (path: string, lang: string) => {
        return mockTableResult;
    },
}));

describe('Rendering test', () => {
    it('renders correctly', () => {
        const { asFragment } = render(<MemoryRouter><UiLanguageContext.Provider value={{ language, setLanguage, languageTab, setLanguageTab, availableUiLanguages, uiContentLanguage, setUiContentLanguage }}><TableListSelection /></UiLanguageContext.Provider></MemoryRouter>);
        expect(asFragment()).toMatchSnapshot();
    });
});

describe('Assertion tests', () => {
    it('should show the contents with primary database language if given UI language is not supported', () => {
        const { asFragment } = render(
            <MemoryRouter>
                <UiLanguageContext.Provider value={{ language: "fr", setLanguage, languageTab, setLanguageTab, availableUiLanguages, uiContentLanguage, setUiContentLanguage }}>
                    <TableListSelection />
                </UiLanguageContext.Provider>
            </MemoryRouter>);
        expect(asFragment().textContent).toContain('foo1-fi');
    });
});
