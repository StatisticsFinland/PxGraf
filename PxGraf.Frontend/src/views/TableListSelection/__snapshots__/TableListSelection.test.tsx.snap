// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering test renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthMd sc-hOTSeL jknTDC css-xdmu94-MuiContainer-root"
  >
    <ul
      class="MuiList-root MuiList-padding css-822hb5-MuiList-root"
    >
      <a
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
        href="/table-list/f/o/o///b/a/r/id1/"
        id="mainContent"
        tabindex="0"
      >
        <div
          class="MuiListItemText-root css-tlelie-MuiListItemText-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-x61gb3-MuiTypography-root"
          >
            foo1-fi
          </p>
        </div>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </a>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
      />
      <a
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
        href="/table-list/f/o/o///b/a/r/id2/"
        id="mainContent"
        tabindex="0"
      >
        <div
          class="MuiListItemText-root css-tlelie-MuiListItemText-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-x61gb3-MuiTypography-root"
          >
            foo2-fi
          </p>
        </div>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </a>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
      />
      <a
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
        href="/editor/foo/bar/id2.px/"
        id="mainContent"
        tabindex="0"
      >
        <div
          class="MuiListItemText-root css-tlelie-MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-x61gb3-MuiTypography-root"
            >
              foo3-fi
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2 css-1wjvmi0-MuiTypography-root"
            >
              tableSelect.updated: 1.1.2000 klo 0.00.00
            </p>
          </span>
        </div>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </a>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
      />
    </ul>
  </div>
</DocumentFragment>
`;
